rust/kernel/rbtree.rs

```diff
@@ -301,6 +313,427 @@ impl<K, V> RBTree<K, V> {
         None
     }
 
+    fn neighborhood_search_raw(
```

Comment:
This helper method might look unwieldy, but I find it better than having to repeat a lot of logic for the 6 types of searches that I support.  Each one of them is just a slight tweak on the underlying binary search algorithm.

I supported all flavors in the spirit of completeness...

Reply:
How about implementing the tree search as a iterator?

Reply:
This will be challenging for me, but I am up for assuming it helps speed up the rust binder :)

Is this the idea?
`fn iter_from_key(key: &K, bound: NeighborType) -> RBTreeIterator`, where `bound` means to start at the upper or lower bound if the given key doesn't exist.

It would also need to be a bidirectional iterator (or a cursor-like thing), presumably.

Reply:
My impression is that what would be useful for binder is not an ordinary iterator, but a cursor. The cursor would be an object that has `go_left` and `go_right` methods, and a way of getting the key/value of the current node. We might also want a `peek_left` and `peek_right` method to avoid doing `go_left`,`get`,`go_right` when merging adjacent nodes in the binder range allocator. Similarly, I believe we would want a delete method that deletes the current node and returns a cursor pointing at the left/right neighbour of the deleted node.